AX: [AX Thread Text APIs] Soft line breaks trim "space" from text runs

hoffmanjoshua opened the pull request:
#### ca3a2787246df9e900fe011fb8b504299d6eb8e6
<pre>
AX: [AX Thread Text APIs] Soft line breaks trim &quot;space&quot; from text runs
<a href="https://bugs.webkit.org/show_bug.cgi?id=290864">https://bugs.webkit.org/show_bug.cgi?id=290864</a>
<a href="https://rdar.apple.com/problem/148366358">rdar://problem/148366358</a>

Reviewed by Tyler Wilcock.

Prior to this point, AXTextRun(s) were not aware of trimmed spaces that occur when rendering soft line
breaks. This means that when text is rendered like the following (as a result of width constraints)

```
Hello
world
```

and we get a text marker range that spans across this line break, we would return a string of &quot;Helloworld&quot;,
rather than &quot;Hello world&quot;, which is what happens on the live tree.

To solve this problem, we re-add these trimmed characters when constructing the AXTextRuns, which is done
by looking for any gaps in the DOM offsets. By doing so, we now properly consider this character when
computing ranges, indices of text markers, and other APIs.

This change also creates new behavior. Now, the space character that isn&apos;t actually rendered is now exposed
to ATs. This means, as an example, users can navigate by character to it. This character won&apos;t have any
width, since it isn&apos;t rendered, but is included in the string and ranges we serve.

Two tests were added to verify this change. Additionally, two tests were updated to &quot;trim&quot; the new space
character we expose at the end of lines, so that the live and isolated trees can share the same test
expectations. Our new tests will provide the coverage gap that this creates.

* LayoutTests/accessibility/isolated-tree/soft-line-breaks-include-space-expected.txt: Added.
* LayoutTests/accessibility/isolated-tree/soft-line-breaks-include-space.html: Copied from LayoutTests/accessibility/mac/line-range-at-soft-breaks.html.
* LayoutTests/accessibility/mac/line-range-at-soft-breaks-expected.txt:
* LayoutTests/accessibility/mac/line-range-at-soft-breaks.html:
* LayoutTests/accessibility/mac/line-requests-starting-after-first-line-expected.txt:
* LayoutTests/accessibility/mac/line-requests-starting-after-first-line.html:
* LayoutTests/accessibility/mac/text-marker-range-across-soft-break-includes-space-expected.txt: Added.
* LayoutTests/accessibility/mac/text-marker-range-across-soft-break-includes-space.html: Added.
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::AXTextRuns):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::AXTextRuns): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):

Canonical link: <a href="https://commits.webkit.org/293140@main">https://commits.webkit.org/293140@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b396767952ab0971bb89cfe0e88b7ad7001f193

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97970 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17595 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103086 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48500 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100015 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17889 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26048 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74596 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31783 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100973 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13554 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88551 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54956 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13339 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6481 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47942 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83325 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6564 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105464 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25050 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18253 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83586 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25422 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84729 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83035 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20975 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27676 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5364 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18678 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25011 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30190 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24833 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26407 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1498,6 +1497,9 @@ AXTextRuns AccessibilityRenderObject::textRuns()
         currentLineIndex = newLineIndex;
 
+        // Within each iteration of this loop, we are looking at the *next* text box to compare to the current.
+        // So, we need to set the textRunDomOffsets after the line index comparison, in order to assign the right DOM offsets per Text Box.
```
"Text Box" is capitalized here, but not in the previous sentence. The uncapitalized version is probably right?

minorninth commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -43,6 +43,26 @@
+        if (!runs[i].text.containsOnlyASCII())
+            containsOnlyASCII = false;
+
+        // If a space was trimmed in this text run (i.e., there's a gap between the end of the current run's DOM offset and the start of the next), add it back.
```
The fix is super clean, just curious if you considered putting this logic in the function that constructs AXTextRuns rather than here, or some other design. I feel like normally we don't do that sort of logic in a constructor.

I wonder if just a helper function to fix up the text runs before calling the constructor might be more clean.

hoffmanjoshua replied:
The first method I considered was doing this when we construct each individual `AXTextRun`, but the issue there is that we need the next text run's DOM offsets to compute this properly. 

Per your suggestion, we can definitely loop through the text runs before calling the constructor and update the text there. I figured since we are already looping here in the constructor, we can tack on to that, but open to either! What do you think?

minorninth reviewed (commented):
lgtm with one suggestion

hoffmanjoshua committed:
AX: [AX Thread Text APIs] Soft line breaks trim "space" from text runs
https://bugs.webkit.org/show_bug.cgi?id=290864
rdar://problem/148366358

Reviewed by Tyler Wilcock.

Prior to this point, AXTextRun(s) were not aware of trimmed spaces that occur when rendering soft line
breaks. This means that when text is rendered like the following (as a result of width constraints)

```
Hello
world
```

and we get a text marker range that spans across this line break, we would return a string of "Helloworld",
rather than "Hello world", which is what happens on the live tree.

To solve this problem, we re-add these trimmed characters when constructing the AXTextRuns, which is done
by looking for any gaps in the DOM offsets. By doing so, we now properly consider this character when
computing ranges, indices of text markers, and other APIs.

This change also creates new behavior. Now, the space character that isn't actually rendered is now exposed
to ATs. This means, as an example, users can navigate by character to it. This character won't have any
width, since it isn't rendered, but is included in the string and ranges we serve.

Two tests were added to verify this change. Additionally, two tests were updated to "trim" the new space
character we expose at the end of lines, so that the live and isolated trees can share the same test
expectations. Our new tests will provide the coverage gap that this creates.

* LayoutTests/accessibility/isolated-tree/soft-line-breaks-include-space-expected.txt: Added.
* LayoutTests/accessibility/isolated-tree/soft-line-breaks-include-space.html: Copied from LayoutTests/accessibility/mac/line-range-at-soft-breaks.html.
* LayoutTests/accessibility/mac/line-range-at-soft-breaks-expected.txt:
* LayoutTests/accessibility/mac/line-range-at-soft-breaks.html:
* LayoutTests/accessibility/mac/line-requests-starting-after-first-line-expected.txt:
* LayoutTests/accessibility/mac/line-requests-starting-after-first-line.html:
* LayoutTests/accessibility/mac/text-marker-range-across-soft-break-includes-space-expected.txt: Added.
* LayoutTests/accessibility/mac/text-marker-range-across-soft-break-includes-space.html: Added.
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::AXTextRuns):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::AXTextRuns): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):

Canonical link: https://commits.webkit.org/293140@main
